feat: integrate frontend + add Review Code Conduct api call

manvinderjit opened the pull request:
add api call to Review Conduct Code page for signing documents.

## Summary

Integrate the frontend pages for the profile completion flow.

## Tests

Requires setting up e2e testing

## Additional information

Closes #223

manvinderjit committed:
feat: add Review Code Conduct api call

add api call to Review Conduct Code page for signing documents.

manvinderjit committed:
fix: useProfileRedirect prevent state updates

update upProfileRedirect hook to prevent state updates on unmounted components.
uses React's idiomatic signal.aborted check instead of a isMounted state.

manvinderjit committed:
fix: useProfileRedirect allows same status for multiple pages

fix loop bug that redirected users on ReviewConductCode to CheckSteps.
update the useProfileRedirect hook to support same status for multiple pages.

manvinderjit committed:
refactor: update CheckSteps to use useProfileRedirect hook

manvinderjit committed:
refactor: update ReviewConductCode to use useProfileRedirect hook

manvinderjit committed:
refactor: update CompleteProfile to use useProfileRedirect hook

refactor CompleteProfile to use userProfileRedirect hook.
fix type in state name setIsSubmission Disabled.
render form based on the loading state.
disable button when loading.

manvinderjit committed:
refactor: CompleteProfile add state for form submission

add separate state for from submission, instead of using loading
in CompleteProfile component.

manvinderjit commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -316,353 +316,353 @@
+								<div className='details-content-file-upload'>
 									{photoFile && (
+										<picture>
+											<img src={photoFile} />
```
will fix when integrating the form with file upload api.

Created an issue here for this: #239

madcampos commented on src/components/CheckSteps/CheckSteps.tsx:
```diff
@@ -1,31 +1,40 @@
+		);
+	}
+
+	return <div>loading...</div>;
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Instead of doing a big if statement and then returning the fallback. Return the fallback early, it makes the code easier to understand, easier to read and is less work for the interpreter to parse and analyze everything before returning.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -316,353 +317,353 @@ const CompleteProfile = () => {
-											<option value='30'>30</option>
-											<option value='31'>31</option>
-										</select>
+	if (redirectionComplete && !isLoading) {
```
Same thing here: put the loading first as an early return.

madcampos commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -27,14 +27,20 @@ const isValidStatus = (status: string): status is ProfileStatus => VALID_STATUSE
 };
 
-const getRedirectPathForStatus = (status: ProfileStatus): string => {
+const normalizePath = (path: string) => path.replace(/\/+$/u, '');
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** What is this regex trying to do? Please add a comment exlaining it.

kbventures replied:
It's noted already,  "Normalizes a path by removing any trailing slashes (/).",  but it is n ot clear that is precisely referring to what the Regex is doing.

madcampos replied:
As a suggestion: set the regex to a constant and comment that constant.

Please note though that [global regexes update the `lastIndex` property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/global#:~:text=Each%20call%20to%20exec()%20will%20update%20its%20lastIndex%20property%2C%20so%20that%20the%20next%20call%20to%20exec()%20will%20start%20at%20the%20next%20character.) and start from that, so you need to clone the regex if you are declaring it in a global scope but using it multiple times in a local scope.

E.g.:
```typescript
const GLOBAL_REGEX = /a/gu; // <- Regex is global!

function wrongMatch() {
  console.log(GLOBAL_REGEX.exec('aa')); // <- Will match correctly.
  console.log(GLOBAL_REGEX.exec('ab')); // <- Will not match correctly due to lastIndex being updated.
}

function correctMatch() {
  console.log(new RegExp(GLOBAL_REGEX).exec('aa')); // <- Will match correctly.
  console.log(new RegExp(GLOBAL_REGEX).exec('ab')); // <- Will also match correctly because the regex was cloned.
}
```

This can get confusing and annoying, so as a rule of thumb, always clone a regex.

manvinderjit replied:
@madcampos Why not just declare a non-global regex? Is there any advantage of using the 'g' flag?

manvinderjit replied:
Updated based on suggestions!

manvinderjit committed:
fix: add early return to CheckSteps

manvinderjit committed:
fix: add early return to ReviewConductCode

manvinderjit committed:
fix: add early return to CompleteProfile

manvinderjit committed:
fix: add comments for path normalization regex

manvinderjit committed:
fix: add catch block to handleSubmit function

manvinderjit committed:
fix: slack handle validation to tackle the check triggering state update

manvinderjit committed:
fix: remove selected from select element option in CompleteProfile component

kbventures commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -87,29 +102,29 @@ export const useProfileRedirect = () => {
 		const redirect = async () => {
 			try {
-				const redirectPath = await getRedirectPath(signal);
+				// Normalize URL path
```
As we already have a normalize function. It makes this comment confusing.  This does not normalize the URL.  We are only extracting the path.

bcha92 replied:
^same as Marco above: comment here should be `// Extract URL path`

manvinderjit replied:
Technically, it extracts the pathname, removing the domain and special chars like `.` & `..` , `%x` chars. 

But I will simplify it for clarity.

kbventures commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -87,29 +102,29 @@ export const useProfileRedirect = () => {
+				const currentPath = new URL(window.location.href).pathname;
+
+				const redirectPath = await getRedirectPath(signal, currentPath);
+
```
Why are we checking for a signal abort here?  The signal is aborted in the callback clean for useEffect, which happens after.

madcampos replied:
I think the abort signal is for canceling the promise.

https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal

kbventures replied:
@manvinderjit  The abort signal can be triggered in several ways, but in all cases, it must be explicitly called in your code. AbortSignal doesn't abort automatically.

If the only abort signal in your component comes from the useEffect cleanup function, then you should remove the abort signal checks.

manvinderjit replied:
https://github.com/torontojs/communityhub/pull/238#discussion_r2308810239

manvinderjit replied:
@kbventures We are not checking for abort signal in line 114. We are passing the signal to the function so that it can be associated with the fetch api. 

<img width="991" height="162" alt="image" src="https://github.com/user-attachments/assets/ffff1154-1138-4190-a43c-78983c164eea" />

On a side note, I have reverted my hook to an older version. It had a bug, the community helped me figure it out.

kbventures commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -87,29 +102,29 @@ export const useProfileRedirect = () => {
-				// Avoid redirect if already on the correct path
-				if (currentPath !== redirectPath) {
+				// Redirect only if current url path is different from the redirectpath
+				if (currentPath && currentPath !== redirectPath) {
```
currentPath twice?

kbventures commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -68,7 +83,7 @@ export const getRedirectPath = async (signal?: AbortSignal): Promise<RedirectPat
 
 		// Redirects based on profile status or to sign-in if not profile status is returned
-		return (data?.status && isValidStatus(data.status)) ? getRedirectPathForStatus(data.status) : REDIRECT_PATHS.signIn;
+		return (data?.status && isValidStatus(data.status)) ? getRedirectPathForStatus(data.status, currentPath) : REDIRECT_PATHS.signIn;
```
Question: Why do we have an explicit null at line 92?

madcampos replied:
On that note, prefer:
- Not returning anything, as functions by default return `undefined`;
- An empty return to stop function execution (`return;`), it will also implicitly return `undefined`;
- Explicitly return `udnefined` if that is what you are looking for.

Avoid using `null` as it is a weirder value, if you can, always use `undefined`.

manvinderjit replied:
I have updated the same to throw an error now.

kbventures commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -87,29 +102,29 @@ export const useProfileRedirect = () => {
+
+				const redirectPath = await getRedirectPath(signal, currentPath);
+
+				if (signal.aborted || hasRedirected.current) {
```
Same comment as above.  The signal is only aborted in the cleanup handler for useEffect.

manvinderjit replied:
@kbventures We abort the signal using this:
`controller.abort();`

`signal.aborted `checks if that particular signal was aborted. It only returns true or false. It doesn't abort the signal => it is not a function.

signal.abort is used like an `isComponentMounted` state

manvinderjit replied:
I have reverted my hook to an older version. It had a bug, the community helped me figure it out.

kbventures reviewed (changes requested):
Great work.  I left a few comments and questions.

Please introduce new API tests for these functionalities.

bcha92 commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -316,8 +325,8 @@ const CompleteProfile = () => {
-	if (isLoading) {
-		return <p>Loading...</p>;
+	if (!redirectionComplete || isLoading) {
+		return <div>loading...</div>;
```
**Nitpick**: Is using a `<div>` for a simple text here appropriate? Can we add a `<p>` text element here.

If this is just for a placeholder for something else (i.e. Design team ideas), then it's fine as is.

manvinderjit replied:
We do need a loading component from the design team, but I will still fix this. Good catch.

bcha92 commented on src/components/ReviewConductCode/ReviewConductCode.tsx:
```diff
@@ -7,9 +7,60 @@ import HeaderTitle from '../HeaderTitle/HeaderTitle.tsx';
+
 	const [agreementChecked, setAgreementChecked] = useState(false);
+	const [isSubmitting, setIsSubmitting] = useState(false);
+	// TODO: Display success and error messages
```
**Nitpick**: Is this still a work in progress or something we can kick this down the road (create new issue)?

kbventures replied:
@caducarvalho is working on this, and yes, having a Github issue link
 here would be great.

manvinderjit replied:
Created. https://github.com/torontojs/communityhub/issues/252

bcha92 commented on src/components/ReviewConductCode/ReviewConductCode.tsx:
```diff
@@ -7,9 +7,60 @@ import HeaderTitle from '../HeaderTitle/HeaderTitle.tsx';
+					console.log('All documents signed:', results);
+					window.location.href = '/pages/complete-profile';
+				} else {
+					setSubmitError('Some documents failed to sign. Please try again!');
```
**Non-Blocker:** Should organizers be notified if volunteers fail to sign, or "decline" to sign the Conduct Code of review?

kbventures replied:
@bcha92 

It's not optional.

<img width="1010" height="883" alt="image" src="https://github.com/user-attachments/assets/bfcf03bf-95ba-47ac-bce1-a28773e5c379" />

bcha92 replied:
@kbventures my apologies for the lack of clarity. I actually meant would there be a fallback mechanism (or time limit) if the volunteer did not take any action (basically "do nothing") to consent to the Code of Conduct within a certain time period after registration?

Or does this fall under the wider scope of "registered but not confirmed" users, similar conditions if a user did not verify their email address after registration?

manvinderjit replied:
@bcha92 Operationally, the user will be redirected to this page every time the sign in, if they don't accept. 

There was a discussion on imposing a limit and deactivating the accounts if they don't accept CoC, after 90 or 180 days, but nothing is concrete yet. @kbventures Might have more on this.

manvinderjit replied:
Resolving this for now. We can bring this up in a future developer meeting.

bcha92 reviewed (commented):
Looks good, but given we now have some clarity (from our meeting today) on how to move forward with the `signal.abort`, we can wait for the changes before it looks good to go.

madcampos commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -51,7 +66,7 @@ const getRedirectPathForStatus = (status: ProfileStatus): string => {
 };
 
-export const getRedirectPath = async (signal?: AbortSignal): Promise<RedirectPathResult> => {
+export const getRedirectPath = async (signal?: AbortSignal, currentPath?: string): Promise<RedirectPathResult> => {
```
**nitpick**: The order of the parameters should be reversed for better UX.

Imagine the most common scenario for using this function, which parameter will be most probable to be passed, the path or the abort signal?

manvinderjit replied:
@madcampos The currentPath is only passed to a sub-function `getRedirectPathForStatus` where it is used to determine between `checkSteps` and `Review Conduct Code` as both require the same status 'activated'. Otherwise, Review Conduct Code kept on redirecting to CheckSteps, even though it required the same status. A quick fix made more sense here.

Signal is going to be used for each fetch api request in the function itself, whereas currentPath is just passed down for an odd case.

manvinderjit replied:
The second parameter is optional, so the lint rule won't allow it to be placed first.

manvinderjit committed:
fix: Improve comment for pathname

manvinderjit committed:
fix: use appropriate html tag for loading element

manvinderjit committed:
refactor: useProfileRedirect hook for simpler functionality

manvinderjit committed:
fix: update CheckSteps for hook refactor

update the CheckSteps component to align with useProfileRedirect hook refactoring.

manvinderjit committed:
fix: update ReviewConductCode for hook refactor

update the ReviewConductCode component to align with useProfileRedirect hook refactoring.

manvinderjit committed:
fix: update CompleteProfile for hook refactor

update the CompleteProfile component to align with useProfileRedirect hook refactoring.

manvinderjit committed:
refactor: remove trailing-slashes regex as a constant

kbventures reviewed (dismissed):
Please fix this before finishing this pr. 

https://github.com/torontojs/communityhub/issues/256

manvinderjit commented:
> Please fix this before finishing this pr.
> 
> #256

We postponed this for now, as per the last meeting.

manvinderjit committed:
Merge latest changes from dev to profile-integration
